src/util.rs

```diff
@@ -197,6 +198,32 @@ pub(crate) fn validate_bao(
     Ok(())
 }
 
+/// converts a canonicalized relative path to a string, returning an error if
+/// the path is not valid unicode
+///
+/// this will also fail if the path is non canonical, i.e. contains `..` or `.`,
+/// or if the path components contain any windows or unix path separators
+pub fn canonicalize_path(path: impl AsRef<Path>) -> anyhow::Result<String> {
+    let parts = path
+        .as_ref()
+        .components()
+        .map(|c| {
+            let c = if let Component::Normal(x) = c {
+                x.to_str().context("invalid character in path")?
+            } else {
+                anyhow::bail!("invalid path component {:?}", c)
+            };
+            anyhow::ensure!(
+                !c.contains('/') && !c.contains('\\'),
+                "invalid path component {:?}",
+                c
+            );
+            Ok(c)
+        })
+        .collect::<anyhow::Result<Vec<_>>>()?;
+    Ok(parts.join("/"))
+}
```

Comment:
There are now two disparate utilities to handle this, this one and `pathbuf_from_name`.  I wonder if it's better to add a `BlobNameWithPathComponents` (or whatever bikeshedding produces a better name) which does this cannoicalisation on `::new()` and has a `to_pathbuf()` method.